src/D2L.CodeStyle.Analyzers/Language/ParamNamesShouldMatchInterfaceAnalyzer.cs

```diff
@@ -0,0 +1,95 @@
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers.Language {
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	internal sealed class ParamNamesShouldMatchInterfaceAnalyzer : DiagnosticAnalyzer {
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+			Diagnostics.InterfaceImplementationParamNameMismatch
+		);
+
+		public override void Initialize( AnalysisContext context ) {
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction( RegisterAnalyzer );
+		}
+
+		private static void RegisterAnalyzer( CompilationStartAnalysisContext context ) {
+			context.RegisterSyntaxNodeAction(
+				ctx => AnalyzeClassDefinition(
+					ctx,
+					( ClassDeclarationSyntax )ctx.Node,
+					( INamedTypeSymbol )ctx.SemanticModel.GetDeclaredSymbol( ctx.Node )
+				),
+				SyntaxKind.ClassDeclaration
+			);
+		}
+
+		private static void AnalyzeClassDefinition(
+			SyntaxNodeAnalysisContext context,
+			ClassDeclarationSyntax typeNode,
+			INamedTypeSymbol typeSymbol
+		) {
+			if( !typeSymbol.IsDefinition ) {
+				return;
+			}
+
+			if( !typeSymbol.Interfaces.Any() ) {
+				return;
+			}
+
+			ImmutableArray<IMethodSymbol> interfaceMethods = CollectInterfaceMethods( typeSymbol );
+			foreach( var interfaceMethod in interfaceMethods ) {
+				IMethodSymbol classMethod = (IMethodSymbol)typeSymbol.FindImplementationForInterfaceMember( interfaceMethod );
+
+				if( classMethod.Parameters.Length == 0 ) {
+					continue;
+				}
+
+				ImmutableArray<IParameterSymbol> classMethodParameters = classMethod.Parameters;
+				ImmutableArray<IParameterSymbol> interfaceMethodParameters = interfaceMethod.Parameters;
+				for( int i = 0; i < classMethodParameters.Length; ++i ) {
+					string classMethodParameterName = classMethodParameters[ i ].Name;
+					string interfaceMethodParameterName = interfaceMethodParameters[ i ].Name;
+
+					if( classMethodParameterName.Equals( interfaceMethodParameterName, StringComparison.InvariantCultureIgnoreCase ) ) {
+						continue;
+					}
+
+					Location location = classMethodParameters[ i ].DeclaringSyntaxReferences.First().GetSyntax().GetLocation();
+					context.ReportDiagnostic( Diagnostic.Create(
+						Diagnostics.InterfaceImplementationParamNameMismatch,
+						location,
+						interfaceMethodParameterName,
+						classMethodParameterName
+					) );
+				}
+			}
+		}
+
+		private static ImmutableArray<IMethodSymbol> CollectInterfaceMethods(
+			INamedTypeSymbol typeSymbol
+		) {
+			var builder = ImmutableArray.CreateBuilder<IMethodSymbol>();
+
+			foreach( INamedTypeSymbol @interface in typeSymbol.AllInterfaces ) {
+				ImmutableArray<ISymbol> members = @interface.GetMembers();
+				ImmutableArray<IMethodSymbol> methods = members
+					.Where( s => s.Kind == SymbolKind.Method )
+					.Cast<IMethodSymbol>()

```

Comment:
You could maybe do

```cs
.OfType<IMethodSymbol>()
```

rather than

```cs
.Where( s => s.Kind == SymbolKind.Method )
.Cast<IMethodSymbol>()
```

but maybe what you wrote is faster? No preference just thinking out loud